Main.elm

```diff
@@ -36,100 +47,239 @@ type State
+    , Html.Events.onMouseDown address (SetupOrientationField shipId)
+    ]
+    [ Html.text "Horizontal" ]
+  , Html.div
+    [ Html.Attributes.style <| vCheckedStyle :: innerDivStyle
+    , Html.Events.onMouseDown address (SetupOrientationField shipId)
     ]
-  else Html.div [] []
+    [ Html.text "Vertical" ]
+  ]
 
 ---- UPDATE ----
 type Action
   = SetupOrientationField Int
-  | SetupRowField Int String
-  | SetupColumnField Int String
-  | SetupAddShip Int
+  | SetupOrientationToggle
+  | SetupSelectShip (Maybe Int)
+  | SetupShowShip (Maybe (Int, Int))
+  | SetupAddShip ()
   | SetupPlay
+  | PlayAim (Maybe (Int, Int))
+  | PlayShoot ()
 
 update : Action -> Model -> Model
 update action model =
   case action of
     SetupOrientationField shipId ->
       { model | player <- Player.updateShip shipId Ship.toggleOrientation model.player }
-    SetupRowField shipId rowAsString ->
-      let
-      updateRow ship =
-        Ship.setRow (toIntOrDefaultOrZero rowAsString (Ship.getRow ship)) ship
-      in
-      { model | player <- Player.updateShip shipId updateRow model.player }
-    SetupColumnField shipId columnAsString ->
-      let
-      updateColumn ship =
-        Ship.setColumn (toIntOrDefaultOrZero columnAsString (Ship.getColumn ship)) ship
-      in
-      { model | player <- Player.updateShip shipId updateColumn model.player }
-    SetupAddShip shipId ->
-      { model | player <- Player.addShip shipId model.player}
+    SetupOrientationToggle ->
+      case model.selectedShipId of
+        Just id ->
+          { model | player <- Player.updateShip id Ship.toggleOrientation model.player }
+        Nothing ->
+          model
+    SetupSelectShip shipId ->
+      { model | selectedShipId <- shipId }
+    SetupShowShip position ->
+      case model.selectedShipId of
+        Just id ->
+          let -- TODO Put all this horror inside a function
+          oldShip =
+            case Fleet.getShip id model.player.fleet of
+              Just ship ->
+                ship
+              Nothing -> -- Never gonna happen
+                Ship.init 1 Ship.Horizontal (0,0)
+          grid = Grid.hideShip oldShip model.player.fleet model.player.primaryGrid
+          in
+          case position of
+            Just pos ->
+              let
+                newShip =
+                  case Fleet.getShip id model.player.fleet of
+                    Just ship ->
+                      Ship.setLocation pos ship
+                    Nothing -> -- Never gonna happen
+                      Ship.init 1 Ship.Horizontal (0,0)
+                g = Grid.showShip newShip model.player.fleet grid
+                fn s = Ship.setLocation pos s
+              in
+              { model | player <- Player.updateGrid g <| Player.updateShip id fn model.player }
+            Nothing ->
+              { model | player <- Player.updateGrid grid model.player }
+        Nothing ->
+          model
+    SetupAddShip _ ->
```

Comment:
Yeah, just trace it inside the code, and you'll see what I mean. It's sends a void signal — `()`, and it has to go somewhere